# LocusCompare: Visualizing the colocalization of association summary statistics 

---
## Introduction
<img src="logo.png" width="600"/>

LocusCompare is a suite of tools and datasets to visualize the colocalization of summary statistics from pairs of association datasets (e.g. GWAS and eQTL). These tools and datasets include: 

- Colocalization statistics between more than 200 GWAS and 48 GTEx (version 7) tissues
- Interactive LocusCompare plot to visualize a single locus
- Bash script to download summary statistics of more than 200 GWAS

## Documentations
For instructions please follow these links: 


- <a href="https://www.github.com/boxiangliu/locuscompare/wiki" target="_blank">Documentation</a>
- <a href="https://www.github.com/boxiangliu/locuscompare/wiki/FAQ" target="_blank">Frequently Asked Questions</a>
- <a href="https://www.github.com/boxiangliu/locuscompare/issues" target="_blank">Bug reports</a>

To download stand-alone version:

- <a href="https://www.github.com/boxiangliu/locuscomparer" target="_blank">LocusCompareR</a>


## License
GPL v3 or later

## Developers
- Boxiang Liu (author, maintainer)
- Mike Gloudemans (author, maintainer)
- Abhiram Rao (author)
- Stephen Montgomery (author)

## Citation
If you use locuscompare, please cite the following paper: https://www.nature.com/articles/s41588-019-0404-0

Boxiang Liu, Michael J. Gloudemans, Abhiram S. Rao, Erik Ingelsson & Stephen B. Montgomery (2019) Abundant associations with gene expression complicate GWAS follow-up, Nature Genetics
